velox/exec/Window.cpp

```diff
@@ -282,6 +314,131 @@ void Window::callResetPartition(vector_size_t partitionNumber) {
+  auto lastPartitionRow = partitionStartRows_[currentPartition_ + 1] - 1;
+  auto type = windowFrames_[functionIndex].type;
+  auto boundType = isStartBound ? windowFrames_[functionIndex].startType
+                                : windowFrames_[functionIndex].endType;
+  auto frameArg = isStartBound ? windowFrames_[functionIndex].start
+                               : windowFrames_[functionIndex].end;
+
+  switch (boundType) {
+    case core::WindowNode::BoundType::kUnboundedPreceding:
+      std::memset(rawFrameBounds, 0, numRows * sizeof(vector_size_t));
+      break;
+    case core::WindowNode::BoundType::kUnboundedFollowing:
+      std::fill_n(
+          rawFrameBounds, numRows, lastPartitionRow - firstPartitionRow);
+      break;
+    case core::WindowNode::BoundType::kCurrentRow: {
+      if (type == core::WindowNode::WindowType::kRange) {
+        const vector_size_t* rawPeerBuffer =
+            isStartBound ? rawPeerStarts : rawPeerEnds;
+        std::copy(rawPeerBuffer, rawPeerBuffer + numRows, rawFrameBounds);
+      } else {
+        // Fills the frameBound buffer with increasing value of row indices
+        // (corresponding to CURRENT ROW) from the startRow of the current
+        // output buffer. The startRow has to be adjusted relative to the
+        // partition start row.
+        std::iota(
+            rawFrameBounds,
+            rawFrameBounds + numRows,
+            startRow - firstPartitionRow);
+      }
+      break;
+    }
+    case core::WindowNode::BoundType::kPreceding: {
+      if (type == core::WindowNode::WindowType::kRows) {
+        updateKRowsFrameBounds(
+            true, frameArg.value(), startRow, numRows, rawFrameBounds);
+      } else {
+        VELOX_NYI("k preceding frame is only supported in ROWS mode");
+      }
+      break;
+    }
+    case core::WindowNode::BoundType::kFollowing: {
+      if (type == core::WindowNode::WindowType::kRows) {
+        updateKRowsFrameBounds(
+            false, frameArg.value(), startRow, numRows, rawFrameBounds);
+      } else {
+        VELOX_NYI("k following frame is only supported in ROWS mode");
+      }
+      break;
+    }
+    default:
+      VELOX_USER_FAIL("Invalid frame bound type");
+  }
+}
+
+namespace {
+void fixOutOfBoundFrames(
+    const vector_size_t& lastRow,
+    const vector_size_t& numRows,
+    vector_size_t* frameStart,
+    vector_size_t* frameEnd) {
+  vector_size_t* frameStartIndex = nullptr;
+  vector_size_t* frameEndIndex = nullptr;
+
+  for (auto i = 0; i < numRows; i++) {
+    frameStartIndex = frameStart + i;
+    frameEndIndex = frameEnd + i;
+    // Clamp frameStart and frameEnd to the first and last row of the partition
+    // if required.
+    if (*frameStartIndex <= *frameEndIndex) {
+      *frameStartIndex = std::max(*frameStartIndex, 0);
+      *frameEndIndex = std::min(*frameEndIndex, lastRow);
+    } else {
+      VELOX_NYI("Empty frames are currently not supported");
```

Comment:
Why are empty frames not supported?

Reply:
@mbasmanova : Empty frames result in null output values from the Window function. The functions code need to be changed to handle this. There are several code changes for it. So it seemed better to limit the changes in this PR.